libs-scala/resources-akka/src/main/scala/com/digitalasset/resources/akka/BoundedSourceQueueResourceOwner.scala

```diff
@@ -9,13 +9,16 @@ import com.daml.resources.{AbstractResourceOwner, HasExecutionContext, Releasabl
 
 import scala.concurrent.Future
 
-class BoundedSourceQueueResourceOwner[T, Context: HasExecutionContext](
-    queueGraph: RunnableGraph[BoundedSourceQueue[T]]
+class BoundedSourceQueueResourceOwner[T, E, Context: HasExecutionContext](
+    queueGraph: RunnableGraph[T],
+    toSourceQueue: T => BoundedSourceQueue[E],
+    toDone: T => Future[Unit],
 )(implicit
     materializer: Materializer
-) extends AbstractResourceOwner[Context, BoundedSourceQueue[T]] {
-  override def acquire()(implicit context: Context): Resource[Context, BoundedSourceQueue[T]] =
-    ReleasableResource(Future(queueGraph.run())) { queue =>
-      Future(queue.complete())
+) extends AbstractResourceOwner[Context, T] {
+  override def acquire()(implicit context: Context): Resource[Context, T] =
+    ReleasableResource(Future(queueGraph.run())) { value =>
+      toSourceQueue(value).complete()
+      toDone(value)
```

Comment:
I also don't get why they didn't expose `watchForCompletion`, as it would make it easier. To force using the materialized value you think?

Reply:
Yeah, I think so. You do have all the information. And just because the queue is complete doesn't mean downstream has finished processing, so this is probably safer.